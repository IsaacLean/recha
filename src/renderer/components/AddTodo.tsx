import MaterialIcon from '@material/react-material-icon'
import React, { Component } from 'react'
import TextField, { Input } from '@material/react-text-field'

interface State {
  value: string
}

export default class AddTodo extends Component<{}, State> {
  public state = {
    value: ''
  }

  private handleChange = evt => {
    this.setState({ value: evt.target.value })
  }

  public render(): JSX.Element {
    const { value } = this.state

    let carriageReturn
    if (value) carriageReturn = <MaterialIcon icon="subdirectory_arrow_left" />

    return (
      <TextField label="New Todo" leadingIcon={<MaterialIcon icon="add" />} trailingIcon={carriageReturn}>
        <Input value={value} onChange={this.handleChange} />
      </TextField>
    )
  }
}
