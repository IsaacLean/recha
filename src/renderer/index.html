<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <%= csp %>
    <title>Todo List</title>
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>