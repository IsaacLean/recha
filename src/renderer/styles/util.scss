.util-cursor-grab {
  cursor: grab !important;
}

.util-cursor-pointer {
  cursor: pointer !important;
}

.util-display-block {
  display: block !important;
}

.util-display-none {
  display: none !important;
}
