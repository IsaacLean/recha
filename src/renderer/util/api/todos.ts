/* eslint-disable import/prefer-default-export */
import fetch from 'cross-fetch'

import HTTPErr from '../HTTPErr'
import Todo from '../../types/Todo'
import { APIRes } from '../../types'
import { genReqOptions } from '.'

interface TodosQuery {
  col: 'id' | 'date' | 'order_num'
  date: string
  dir: 'ASC' | 'asc' | 'DESC' | 'desc'
}

// @ts-ignore
const TODOS_PATH = `${__API__}/todos` // eslint-disable-line no-undef

export const deleteTodo = (id: number): Promise<APIRes<Todo>> =>
  fetch(`${TODOS_PATH}/${id}`, genReqOptions('DELETE')).then(res => {
    if (!res.ok) throw new HTTPErr(res.statusText, res.status, res)
    return res.json()
  })

export const getTodoLists = (): Promise<APIRes<string[]>> =>
  fetch(`${TODOS_PATH}/lists`).then(res => {
    if (!res.ok) throw new HTTPErr(res.statusText, res.status, res)
    return res.json()
  })

export const getTodos = (query?: Partial<TodosQuery>): Promise<APIRes<Todo[]>> => {
  let newPathTodos = TODOS_PATH
  if (query) {
    const queryString = new URLSearchParams(query).toString()
    if (queryString) newPathTodos += `?${queryString}`
  }

  return fetch(newPathTodos).then(res => {
    if (!res.ok) throw new HTTPErr(res.statusText, res.status, res)
    return res.json()
  })
}

export const patchTodo = (id: number, data: Partial<Todo>): Promise<APIRes<Todo>> =>
  fetch(`${TODOS_PATH}/${id}`, genReqOptions('PATCH', data)).then(res => {
    if (!res.ok) throw new HTTPErr(res.statusText, res.status, res)
    return res.json()
  })

export const patchTodoOrders = (data: number[]): Promise<APIRes<Todo[]>> =>
  fetch(`${TODOS_PATH}/reorder`, genReqOptions('PATCH', data)).then(res => {
    if (!res.ok) throw new HTTPErr(res.statusText, res.status, res)
    return res.json()
  })

export const postTodo = (data: Partial<Todo>): Promise<APIRes<Todo>> =>
  fetch(`${TODOS_PATH}`, genReqOptions('POST', data)).then(res => {
    if (!res.ok) throw new HTTPErr(res.statusText, res.status, res)
    return res.json()
  })
