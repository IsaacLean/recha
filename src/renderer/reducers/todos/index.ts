import { combineReducers } from 'redux'

import api from './api'
import ui from './ui'

const todos = combineReducers({ api, ui })

export default todos
