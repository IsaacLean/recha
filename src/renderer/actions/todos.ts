import { normalize } from 'normalizr'

import Todo from '../types/Todo'
import { APIRes } from '../types'
import { arrayOfTodos } from './schema'
import { getTodos } from '../util/api'
import {
  ActionFetchTodosFailure,
  ActionFetchTodosReq,
  ActionFetchTodosSuccess,
  ActionSelectTodo,
  ActionUpdateFormTodoName,
  FETCH_TODOS_FAILURE,
  FETCH_TODOS_REQ,
  FETCH_TODOS_SUCCESS,
  SELECT_TODO,
  UPDATE_FORM_TODO_NAME,
  NormalizedTodosRes,
  ThunkDispatch,
  ThunkResult
} from '../types/actions'
import { State } from '../types/reducers'

export const fetchTodosFailure = (message: string = 'Something went wrong.'): ActionFetchTodosFailure => ({
  type: FETCH_TODOS_FAILURE,
  message
})

export const fetchTodosReq = (): ActionFetchTodosReq => ({
  type: FETCH_TODOS_REQ
})

export const fetchTodosSuccess = (date: Todo['date'], res: APIRes<Todo[]>): ActionFetchTodosSuccess => {
  return {
    type: FETCH_TODOS_SUCCESS,
    date,
    res: normalize(res.data, arrayOfTodos)
  }
}

export const fetchTodos = (date: Todo['date']): ThunkResult<Promise<ActionFetchTodosFailure | NormalizedTodosRes>> => (
  dispatch: ThunkDispatch,
  getState: () => State
) => {
  // if (getIsFetching(getState(), filter)) {
  //   return Promise.resolve()
  // }

  dispatch(fetchTodosReq())

  return getTodos({ date, col: 'order_num', dir: 'ASC' }).then(
    res => Promise.resolve(dispatch(fetchTodosSuccess(date, res)).res),
    err => Promise.resolve(dispatch(fetchTodosFailure(err.message)))
  )
}

export const selectTodo = (id: Todo['id'], name: Todo['name']): ActionSelectTodo => ({
  type: SELECT_TODO,
  id,
  name
})

export const updateFormTodoName = (id: Todo['id'], name: Todo['name']): ActionUpdateFormTodoName => ({
  type: UPDATE_FORM_TODO_NAME,
  id,
  name
})
