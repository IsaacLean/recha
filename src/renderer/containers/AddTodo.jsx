// @flow
import React from 'react'
import { connect } from 'react-redux'

import { addTodo } from '../actions'
import type { Dispatch } from '../types'

let AddTodo = ({ dispatch }: { dispatch: Dispatch }) => {
  let input

  return <>
    <input ref={node => {
      input = node
    }} />
    <button onClick={() => {
      if(input) {
        dispatch(addTodo(input.value))
        input.value = ''
      }
    }}>
      Add Todo
    </button>
  </>
}

AddTodo = connect()(AddTodo)

export default AddTodo
