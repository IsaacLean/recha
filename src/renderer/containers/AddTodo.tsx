import React, { ReactNode } from 'react'
import { connect } from 'react-redux'

import { addTodo } from '../actions'
import { ThunkDispatch } from '../types'

export const AddTodo = ({ dispatch }: { dispatch: ThunkDispatch }): JSX.Element => {
  let input

  return (
    <>
      <input
        ref={node => {
          input = node
        }}
      />
      <button
        onClick={() => {
          if (input) {
            dispatch(addTodo(input.value))
            input.value = ''
          }
        }}
      >
        Add Todo
      </button>
    </>
  )
}

export default connect()(AddTodo)
