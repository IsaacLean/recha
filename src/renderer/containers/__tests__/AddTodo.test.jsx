import React from 'react'
import renderer from 'react-test-renderer'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux'

import AddTodo from '../AddTodo'
import todos from '../../reducers'
import { setupStore } from '../../util/store'

describe('AddTodo', () => {
  it('renders properly', () => {
    const container = renderer.create(
      <Provider store={setupStore(todos, null, [thunk])}>
        <AddTodo />
      </Provider>
    )
    expect(container).toMatchSnapshot()
  })
})
