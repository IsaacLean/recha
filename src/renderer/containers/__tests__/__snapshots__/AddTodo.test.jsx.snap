// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodo renders properly 1`] = `
Array [
  <input />,
  <button
    onClick={[Function]}
  >
    Add Todo
  </button>,
]
`;
